src/mgr/QueryParams.cc

```diff
@@ -68,60 +69,61 @@ Mgr::QueryParams::find(const String& name) const
     return iter;
 }
 
-bool
-Mgr::QueryParams::ParseParam(SBuf &paramStr, Param &param)
+/// Parses the value part of a "param=value" URL section.
+/// Value can be a comma separated list of integers, or an opaque string.
+Mgr::QueryParam::Pointer
+ParseParamValue(SBuf &valueStr)
 {
-    static const CharacterSet nameChars = CharacterSet("param-name", "_") + CharacterSet::ALPHA + CharacterSet::DIGIT;
-
-    Parser::Tokenizer tok(paramStr);
-
-    SBuf name;
-    if (!tok.prefix(name, nameChars))
-        return false;
-
-    if (!tok.skip('='))
-        return false; // TODO: support name-only parameters
-
-    static const CharacterSet badValueChars = CharacterSet("param-value-bad", "&= ");
-    SBuf valueStr = tok.remaining();
-    if (valueStr.findFirstOf(badValueChars) != SBuf::npos)
-        return false;
-
-    param.first = SBufToString(name);
+    static const CharacterSet comma("comma",",");
 
+    Parser::Tokenizer tok(valueStr);
     std::vector<int> array;
     int64_t intVal = 0;
     while (tok.int64(intVal, 10, false)) {
-        static const CharacterSet comma("comma",",");
         (void)tok.skipAll(comma);
         Must(intVal >= std::numeric_limits<int>::min());
         Must(intVal <= std::numeric_limits<int>::max());
-        array.emplace_back(int(intVal));
+        array.emplace_back(intVal);
     }
 
     if (tok.atEnd())
-        param.second = new IntParam(array);
+        return new Mgr::IntParam(array);
     else
-        param.second = new StringParam(SBufToString(valueStr));
-
-    return true;
+        return new Mgr::StringParam(SBufToString(valueStr));
 }
 
-bool
+/**
+ * Syntax:
+ *   query  = param *( '&' param ) [ '#' fragment ]
```

Comment:
This query parameters-parsing function implies that the fragment is a part of the query but does not consume it like the other documented parts of the query. Since we are sharing the tokenizer with the caller, we are responsible for leaving the tokenizer in the right state upon successful function termination.

I see two ways to fix these problems:

1. Keep the fragment in the function description. Adjust the function implementation to consume the fragment. I suspect that this option should be _rejected_ because fragments may follow well-formed parameterless URLs (i.e. URLs without `?`) that this function does not see. Thus, our caller would have to consume fragments as well. That would needlessly duplicate code and will force us to add Tokenizer::skipEverything() or equivalent (see the note below). I see no need for such complications.
2. Remove the fragment section from the current function interface description. Do not consume `#` and the following fragment in this function. Let the caller handle the fragment, with or without the query. I would try implementing this option first.

----

A note for the code that will consume the fragment if it turns our that such code is needed (I hope it is not): AFAICT, Tokenizer does not have the right method for skipping everything. We should either add Tokenizer::skipEverything() OR, if you prefer, overload skipAll() by adding a second skipAll() method (this one without parameters). The skipEverything() option may be slightly better because it avoids overloading, but I am not sure.

Reply:
Done in 0dc57a18a